### PR TITLE
[python-package] Add Polars / Arrow C Stream support with categorical handling

### LINKED ISSUE #6204: Adding support for polars for input data
## Summary

I think polars library is on the path to replace the majority of pandas use-cases. It is already being adopted by the community. We use it internally in my company for new projects and we try not to use pandas at all.

## Motivation

Polars is blazingly fast and it has several times a lower memory footprint. There is no need to use extra memory to convert data into numpy or pandas to be used for training in LightGBM.

## Description

I would like the following to be working, where `data_train` and `data_test` are instances of `pl.DataFrame`
```python
y_train = data_train[col_target]
y_test = data_test[col_target]
X_train = data_train.select(cols_pred)
X_test = data_test.select(cols_pred)

lgb_train = lgb.Dataset(X_train, y_train)
lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)

params = {
    "boosting_type": "gbdt",
    "objective": "regression",
    "metric": {"l2", "l1"},
    "learning_rate": 0.1,
    "feature_fraction": 0.9,
    "bagging_fraction": 0.8,
    "bagging_freq": 5,
    "verbose": 0,
    "num_leaves": 42,
    "max_depth": 5,
    "num_iterations": 5000,
    "min_data_in_leaf": 500,
    "reg_alpha": 2, 
    "reg_lambda": 5,
}

gbm = lgb.train(
    params,
    lgb_train,
    valid_sets=lgb_eval,
    callbacks=[lgb.early_stopping(stopping_rounds=500)],
)

```
as of now I have to convert it into numpy matrices
```python
y_train = data_train[col_target].to_numpy()
y_test = data_test[col_target].to_numpy()
X_train = data_train.select(cols_pred).to_numpy()
X_test = data_test.select(cols_pred).to_numpy()
```

### LINKED ISSUE #6849: Fitting on Polars Dataframe fails due to missing setter for fetures_names_in_
## Description
When trying to fit a lightgbm model using a polars Dataframe as input, the code fails with the attribute error: 
`AttributeError: property 'feature_names_in_' of 'LGBMRegressor' object has no setter
`

Inheriting from the Model and defining a `feature_names_in` property with a setter fixes the issue. 

This error does not occur when using a pandas DataFrame as Input (Version 2.2.2). 
## Reproducible example
```python
import lightgbm as lgb
import numpy as np
import polars as pl

n = 500
rng = np.random.default_rng(42)

data = {"x1": rng.integers(0, 2, size=n), "x2": rng.integers(0, 2, size=n)}

df = pl.DataFrame(data)

y = data["x1"] + data["x2"] + data["x1"] * data["x2"]
y = y + rng.normal(scale=0.01, size=n)


parameters = {
    "learning_rate": 0.1,
    "min_data_in_bin": 1,
    "min_data_in_leaf": 1,
    "num_iterations": 3,
    "num_leaves": 4,
    "verbosity": -1,
}

# This fails with an AttributeError
regressor = lgb.LGBMRegressor(**parameters)
regressor.fit(df, y).predict(df)


# Rerunning with the PatchedRegressor fixes the issue
class PatchedRegressor(lgb.LGBMRegressor):

    @property
    def feature_names_in_(self):
        return self._feature_name

    @feature_names_in_.setter
    def feature_names_in_(self, x):
        self._feature_name = x


regressor = PatchedRegressor(**parameters)
regressor.fit(df, y).predict(df)
 
```

## Environment info
Lightgbm: 4.6.0
Polars: 1.22.0
Numpy: 2.1.3
Python: 3.11.11

### PR DESCRIPTION
**Pull request by Ic3fr0g**
## Summary

Closes #6204 — adds first-class support for Polars DataFrames (and any other Arrow-compatible object) as input to LightGBM.
Fixes #6849 — `LGBMClassifier/Regressor.fit()` with Polars data no longer raises `AttributeError: property 'feature_names_in_' has no setter`.

---

Changes:
- compat.py: add pa_concat_arrays, pa_RecordBatchReader, arrow_is_dictionary exports and a POLARS_INSTALLED flag
- basic.py: add _has_arrow_c_stream(), _arrow_dict_to_utf8(), _data_from_arrow(); update _lazy_init() and _InnerPredictor.predict() to accept __arrow_c_stream__ objects and handle DictionaryType columns
- sklearn.py: add feature_names_in_ setter (fixes #6849); extend sklearn validation bypass to __arrow_c_stream__ objects in fit() and predict()
- tests/python_package_test/test_polars.py: new tests covering core and sklearn APIs, null categories, unseen categories at inference and PyArrow dict-typed tables

---

Would love your thoughts on this @borchero and @jameslamb and I apologise that it took me [soo long](https://github.com/lightgbm-org/LightGBM/issues/6204#issuecomment-2487390737)

---

Built with help from Claude, reviewed and verified manually.

### CODE REVIEW

**Comment by borchero** [comment]
Thanks for the push @Ic3fr0g! I will get back to this within the next ~week :)

File: tests/python_package_test/test_polars.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+from typing import Any, Dict, Tuple
+
+import numpy as np
+import pytest
+
+import lightgbm as lgb
+from lightgbm.compat import POLARS_INSTALLED, PYARROW_INSTALLED
+
+if not PYARROW_INSTALLED:
+    pytest.skip("pyarrow is not installed", allow_module_level=True)
```
Comment (jameslamb) [inline tests/python_package_test/test_polars.py]: You need to add `polars` to the CI environments, or these tests will never run. Please do that.

Look in `.ci/` for everywhere you see `pandas`, that's probably a good approximation of where you need to list `polars`.

File: tests/python_package_test/test_polars.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+These tests cover every combination of:
+  - data source: Polars DataFrame, PyArrow Table with dictionary columns
+  - API surface: lgb.Dataset + lgb.train / Booster.predict (core) and
+                 LGBMClassifier / LGBMRegressor .fit / .predict (sklearn)
+  - feature types: numeric-only and numeric + categorical
+  - edge cases: null category values, unseen categories at inference,
+                save/load model round-trip
+"""
+
```
Comment (jameslamb) [inline tests/python_package_test/test_polars.py:14]: Any tests that aren't using `polars` and are just generically testing `pyarrow` inputs should move to the existing module for that: https://github.com/lightgbm-org/LightGBM/blob/master/tests/python_package_test/test_arrow.py

I am ok with introducing this new `test_polars.py` module for the `polars` support with whatever `polars`-specific things are left after that. This project could probably benefit from breaking up its Python tests a bit.

Either way, please also look around for other tests checking a wide range of input types and add `polars` inputs where appropriate.

For example, follow wherever these are used:

https://github.com/lightgbm-org/LightGBM/blob/2a675f269d08fd51fad23d67a0214f574acb384e/tests/python_package_test/test_sklearn.py#L66-L68

File: python-package/lightgbm/compat.py [hunk 3]
```diff
@@ -341,6 +352,15 @@ def __init__(self, *args: Any, **kwargs: Any):
...
 
+"""polars"""
+try:
+    import polars as _polars_module  # noqa: F401
+
+    POLARS_INSTALLED = True
+except ImportError:
+    POLARS_INSTALLED = False
+
+
```
Comment (jameslamb) [inline python-package/lightgbm/compat.py]:
```suggestion
```

`POLARS_INSTALLED` appears to only be used in the test code. Please, let's not introduce any new runtime burden for users just for the purpose of test code.

Remove this and use `pytest.importoskip()` in the test code instead (https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-importorskip).

File: python-package/lightgbm/basic.py [hunk 4]
```diff
@@ -1161,6 +1303,18 @@ def predict(
...
+            if _has_arrow_c_stream(data):
+                if not PYARROW_INSTALLED:
+                    raise LightGBMError("Cannot predict from an Arrow C Stream object without 'pyarrow' installed.")
+                data = pa_RecordBatchReader.from_stream(data).read_all()
+            # re-encode any dictionary columns using the saved category maps
+            data = _data_from_arrow(
+                data=data,
+                feature_name="auto",
+                categorical_feature="auto",
+                pandas_categorical=self.pandas_categorical,
```
Comment (jameslamb) [inline python-package/lightgbm/basic.py:1355]: Note to myself (not work for this PR): we should probably introduce a new name for this attribute and deprecate `pandas_categorical`, now that it might be more than just `pandas` where a categorical mapping is stored.

For this PR, proceeding with `pandas_categorical` and tacking on stuff like this here is fine in my opinion.

**Review by jameslamb — CHANGES_REQUESTED** [review]
> _Would love your thoughts on this @borchero and @jameslamb and I apologise that it took me https://github.com/lightgbm-org/LightGBM/issues/6204#issuecomment-2487390737_

Thanks for returning to this!!! I'll defer most of the design discussion to @borchero 's expertise, but I left a few high-level comments.

If this PR reaches a state where @borchero is happy with the interface, I'll provide a more thorough review on smaller structural things, test strictness, etc.

---

When you feel it's close to ready for merging, please take this PR out of draft.

---

Also please note that because you haven't contributed to `lightgbm` before, we'll have to manually approve every CI run. I've done that here so you can get some initial feedback, but I recommend trying to test as much as you can locally.

We need better docs on that, here's how I do it for the Python package:

```shell
# build shared library (does not need to be re-run unless you change C/C++ side)
rm -rf ./build
cmake -B build -S .
cmake --build build --target _lightgbm -j4

# build and install Python package (re-run whenever you make Python changes)
sh build-python.sh install --precompile

# run tests
pytest tests/python_package_test/
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -15,6 +15,70 @@ Your help is very valuable to make it better for everyone.
 
 ## Development Guide
 
+### Python Package Development
```
Comment (jameslamb) [inline CONTRIBUTING.md]: Something like this is fine but please, get it out of this PR. You're welcome to put up a separate PR and if you do, please link it to #6350

Let's please keep this PR focused on its main goal, open separate PRs for other things you discover along the way.

File: tests/python_package_test/test_polars.py (hunk 1 above)
Comment (Ic3fr0g) [reply tests/python_package_test/test_polars.py]: Addressed.

File: tests/python_package_test/test_polars.py (hunk 2 above)
Comment (Ic3fr0g) [reply tests/python_package_test/test_polars.py:14]: Addressed.

File: python-package/lightgbm/compat.py (hunk 3 above)
Comment (Ic3fr0g) [reply python-package/lightgbm/compat.py]: Addressed.

File: CONTRIBUTING.md (hunk 5 above)
Comment (Ic3fr0g) [reply CONTRIBUTING.md]: Addressed.

**Comment by jameslamb** [comment]
Just letting you know, looks like `scikit-learn` adopted `narwhals` in its recent nightlies and that caused a regression in support for `pyarrow` inputs. You may see it show up in tests here.

See https://github.com/lightgbm-org/LightGBM/pull/7268/changes#r3223494907 for a workaround and follow #7269 for a resolution.

**Comment by Ic3fr0g** [comment]
Yeah! I noticed, the workaround I had right now was forcing `y=np.asarray(y)` in `sklearn.py`. Will wait for a proper resolution before moving further on this piece. In the meantime, I've fixed some of the issues in the py39 build.

**Comment by borchero** [comment]
Hey @Ic3fr0g! Once #7275 has been merged, we will be able to simplify this PR significantly:

- There is no need for additional code in `compat.py`
- #7275 adds a dependency on narwhals: categoricals should be identified via narwhals and the categories should be read via [`get_categories`](https://narwhals-dev.github.io/narwhals/api-reference/expr_cat/?h=get_categories#narwhals.expr.ExprCatNamespace.get_categories)
